chat-plugins/modlog.js

```diff
@@ -0,0 +1,273 @@
+	}
+
+	getListClone() {
+		return this.list.slice(0);
+	}
+
+	tryInsert(element) {
+		let insertedAt = -1;
+		for (let i = this.list.length - 1; i >= 0; i--) {
+			if (element.localeCompare(this.list[i]) < 0) {
+				insertedAt = i + 1;
+				if (i === this.list.length - 1) {
+					this.list.push(element);
+					break;
+				}
+				this.list.splice(i + 1, 0, element);
+				break;
+			}
+		}
+		if (insertedAt < 0) this.list.splice(0, 0, element);
+		if (this.list.length > this.maxSize) {
+			this.list.splice(-1, 1);
+			if (insertedAt === this.list.length) return false;
+		}
+		return true;
+	}
+}
+
+function runModlog(room, searchString, exactSearch, maxLines) {
+	let fileNameList = [];
+	if (room === 'all') {
+		const fileList = fs.readdirSync(`${__dirname}/${LOG_PATH}`);
+		for (let i = 0; i < fileList.length; i++) {
+			fileNameList.push(path.normalize(`${__dirname}/${LOG_PATH}${fileList[i]}`));
+		}
+	} else if (room === 'public') {
+		const isPublicRoom = (room => !(room.isPrivate || room.battle || room.isPersonal || room.id === 'global'));
+		const publicRoomIds = Array.from(Rooms.rooms.values()).filter(isPublicRoom).map(room => room.id);
+		for (let i = 0; i < publicRoomIds.length; i++) {
+			fileNameList.push(path.normalize(`${__dirname}/${LOG_PATH}modlog_${publicRoomIds[i]}.txt`));
+		}
+	} else {
+		fileNameList = [path.normalize(`${__dirname}/${LOG_PATH}modlog_${room}.txt`)];
+	}
+
+	let searchStringRegex;
+	if (!searchString) {
+		searchStringRegex = new RegExp('.');
+	} else if (exactSearch) {
+		const escapedSearchString = searchString.replace(/\\/g, '\\\\\\\\').replace(/["'`]/g, '\'\\$&\'').replace(/[\{\}\[\]\(\)\$\^\.\?\+\-\*]/g, '[$&]');
+		searchStringRegex = new RegExp(escapedSearchString);
+		searchString = `"{$searchString}"`;
+	} else {
+		searchString = searchString.replace(/[^a-zA-Z0-9]/, '');
+		searchStringRegex = new RegExp(`\\b${searchString.split('').join('\\W*')}\\b`);
+	}
+
+	let results = new SortedLimitedLengthList(maxLines);
+	for (let i = 0; i < fileNameList.length; i++) {
+		checkRoomModlog(fileNameList[i], searchStringRegex, results);
+	}
+	const resultData = results.getListClone();
+	return resultData.join('\n');
+}
+
+function checkRoomModlog(path, regex, results) {
+	if (!fs.existsSync(path)) return results;
+	let logEntriesTooOld = false;
+	let newLines = [];
+	let lineFragment = '';
+	let blockSize = 1024;
+	let buf = Buffer.alloc(blockSize);
+	const stats = fs.statSync(path);
+	let startPos = stats.size;
+	let fd = fs.openSync(path, 'r');
+
+	do {
+		startPos -= blockSize;
+		if (startPos < 0) {
+			blockSize += startPos;
+			startPos = 0;
+		}
+		let bytesRead = fs.readSync(fd, buf, 0, blockSize, startPos);
+		newLines = buf.toString().substr(0, bytesRead).split('\n').reverse();
+		newLines[0] += lineFragment;
+		lineFragment = newLines[newLines.length - 1];
+		if (startPos > 0) newLines.splice(-1, 1);
+		for (let i = 0; i < newLines.length; i++) {
+			if (newLines[i] && regex.test(newLines[i])) {
```

Comment:
The text before the first newline in a block will be appended to the text after the last newline in the next block and is subsequently treated like any other line, so it should be found just fine.